torch/_inductor/codegen/triton.py

```diff
@@ -1201,6 +1203,26 @@ def __call__(self):
             def _new_line(self, line):
                 return IndirectAssertLine(line, self.var, self.mask, self.size_map)
 
+        if var.bounds.lower < 0:
+            str_size = self.index_to_str(size)
+            new_bounds = ValueRanges.unknown()
+            if var.bounds != ValueRanges.unknown():
+                # Take the negative part of the bound and add size to it
+                # Then take union of that and the positive part
+                # This is a tighter bound than that of a generic ops.where, as we have info on the conde
+                neg = var.bounds & ValueRanges(-sympy.oo, -1)
+                pos = var.bounds & ValueRanges(0, sympy.oo)
+                new_bounds = pos | ValueRanges(neg.lower + size, neg.upper + size)
+
+            new_var = self.cse.generate(
+                self.compute,
+                f"tl.where({var} < 0, {var} + {str_size}, {var})",
+                bounds=new_bounds,
+            )
+
+            new_var.update_on_args("index_wrap", (var,), {})
+            var = new_var
```

Comment:
Should `index_wrap` be it's own operator? `_unsafe_index` for example probably shouldn't generate wrapping.

Reply:
Either a separate operator, or a `wrap=True` argument that we can selectively disable.

Reply:
what about freeride off the `check` argument, and just generate this `check=True`. At the moment we're using `check=False` to circumvent limitations of the bound variable analysis. In all those cases, the indices are always positive so...

Reply:
I can think of examples like backwards functions where you might still need wrapping, but could prove that the indices are in-bounds because they've already been checked elsewhere in the program.

Reply:
Fair enough. Let's write that optimisation in another PR though. I can add the `if check` in this one if you want, though, as it'd be correct with the uses we currently have of `_unsafe_index` and `index_put`.